File: solution-cpp/011_container_with_most_water.cpp
/*
	011. Container With Most Water
	@name: li jin
	@date: Feb 21, 2016
	@link: https://leetcode.com/problems/container-with-most-water/
	@time: 24 ms
*/

class Solution {
public:
    int maxArea(vector<int>& height) {
        int i = 0, j = height.size() - 1;
        int area = 0;
        while (i < j) {
            int lo = height[i];
            int hi = height[j];
            if (lo < hi) {
                area = max(area, (j - i) * height[i]);
                while (i < j && height[i] <= lo) i++;
            }
            else {
                area = max(area, (j - i) * height[j]);
                while (i < j && height[j] <= hi) j--;
            }
        }
        return area;
    }
};

File: solution-python/011_container_with_most_water.py
"""
	011. Container With Most Water
	@name: li jin
	@date: Feb 21, 2016
	@link: https://leetcode.com/problems/container-with-most-water/
	@time: 68 ms
"""

class Solution(object):
    def maxArea(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        i, j = 0, len(height) - 1
        area = 0
        while i < j:
            lo, hi = height[i], height[j]
            if lo < hi:
                area = max(area, (j - i) * lo)
                while i < j and height[i] <= lo:
                    i += 1
            else:
                area = max(area, (j - i) * hi)
                while i < j and height[j] <= hi:
                    j -= 1
        return area
